mrp_account_analytic_wip/models/mrp_production.py

```diff
@@ -418,59 +419,62 @@ def action_confirm(self):
+        self.populate_ref_bom_tracking_items()
         return res
 
-    def _prepare_bom_raw_tracking_items(self, item):
-        analytic = self.analytic_account_id
-        return {
-            "analytic_id": analytic.id,
-            "product_id": item.product_id.id,
-            "planned_qty": item.product_qty,
-        }
-
-    def _create_bom_raw_tracking_items(self, reference_bom_id):
-        """
-        When creating a Raw Material Analytic Item,
-        link it to a BoM Raw Tracking Item, that may have to be created if it doesn't exist.
-        """
-        self._create_bom_tracking_items(
-            reference_bom_id.bom_line_ids, self._prepare_bom_raw_tracking_items
-        )
-
-    def _prepare_bom_ops_tracking_items(self, item):
-        analytic = self.analytic_account_id
-        return {
-            "analytic_id": analytic.id,
-            "product_id": item.workcenter_id.analytic_product_id.id,
-            "planned_qty": item.time_cycle / 60,
-        }
-
-    def _create_bom_ops_tracking_items(self, reference_bom_id):
+    def _populate_ref_bom_tracking_items(self, tracking_item_values):
         """
-        When creating an Operations Analytic Item,
-        link it to a BoM Operations Tracking Item,
-        that may have to be created if it doesn't exist.
+        Creates or updates Tracking Items for the Reference BOM
         """
-        self._create_bom_tracking_items(
-            reference_bom_id.operation_ids, self._prepare_bom_ops_tracking_items
-        )
-
-    def _create_bom_tracking_items(self, items, _prepare_bom_tracking_items):
         self.ensure_one()
-        for item in items:
-            vals = _prepare_bom_tracking_items(item)
-            tracking = self.env["account.analytic.tracking.item"].create(vals)
-            if tracking.product_id not in self.analytic_tracking_item_ids.product_id:
-                self.bom_analytic_tracking_item_ids += tracking
+        # Ensure the MO has an Analytic Account set
+        if not self.analytic_account_id:
+            raise exceptions.UserError(
+                _("Analytic Account is missing for %s.") % self.display_name
+            )
+        # Create missing Tracking Items form BOm components
+        # For each BOM componet value, find a matching Tracking Item,
+        # and update or create it.
+        TrackingItem = self.env["account.analytic.tracking.item"]
+        existing_tracking_items = self.analytic_tracking_item_ids
+        new_tracking_items = TrackingItem
+        for vals in tracking_item_values:
+            # Set MO and Analytic Account on the values
+            vals_mo = {
+                "analytic_id": self.analytic_account_id.id,
+                "production_id": self.id,
+            }
+            vals.update(vals_mo)
+            # Locate existing Tracking Item
+            workcenter_id = vals.get("workcenter_id")
+            product_id = vals.get("product_id")
+            if workcenter_id:
```

Comment:
This needs improvement: we want a WorkCenter item to match ans existing WorkOrder for that same Workcenter.
Working to add this,